supporting_scripts/analysis-of-endpoint-connections/src/main/java/de/tum/cit/endpointanalysis/EndpointAnalyzer.java

```diff
@@ -0,0 +1,115 @@
+package de.tum.cit.endpointanalysis;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class EndpointAnalyzer {
+
+    private static String EndpointAnalysisResultPath = "endpointAnalysisResult.json";
+
+    private static final Logger logger = LoggerFactory.getLogger(EndpointAnalyzer.class);
+
+    public static void main(String[] args) {
+        analyzeEndpoints();
+        printEndpointAnalysisResult();
+    }
+
+    /**
+     * Analyzes endpoints and matches them with REST calls.
+     *
+     * This method reads endpoint and REST call information from JSON files,
+     * compares them to find matching REST calls for each endpoint, and writes
+     * the analysis result to a JSON file. Endpoints without matching REST calls
+     * are also recorded.
+     */
+    private static void analyzeEndpoints() {
+        ObjectMapper mapper = new ObjectMapper();
+
+        try {
+            List<EndpointClassInformation> endpointClasses = mapper.readValue(new File(EndpointParser.ENDPOINT_PARSING_RESULT_PATH),
+                    new TypeReference<List<EndpointClassInformation>>() {
+                    });
+            List<RestCallFileInformation> restCallFiles = mapper.readValue(new File(EndpointParser.REST_CALL_PARSING_RESULT_PATH),
+                    new TypeReference<List<RestCallFileInformation>>() {
+                    });
+
+            List<UsedEndpoints> endpointsAndMatchingRestCalls = new ArrayList<>();
+            List<EndpointInformation> unusedEndpoints = new ArrayList<>();
+
+            for (EndpointClassInformation endpointClass : endpointClasses) {
```

Comment:
can we maybe consider using a less performance heavy approach. I think 4 for loops are heavy:
Would something like this work where you use a map?

```
  Map<String, List<RestCallInformation>> restCallMap = new HashMap<>();
  for (RestCallFileInformation restCallFile : restCallFiles) {
      for (RestCallInformation restCall : restCallFile.getRestCalls()) {
          String restCallURI = restCall.buildComparableRestCallUri();
          restCallMap.computeIfAbsent(restCallURI, k -> new ArrayList<>()).add(restCall);
      }
  }
  
  for (EndpointClassInformation endpointClass : endpointClasses) {
      for (EndpointInformation endpoint : endpointClass.getEndpoints()) {
          List<RestCallInformation> matchingRestCalls = restCallMap.getOrDefault(endpoint.buildComparableEndpointUri(), new ArrayList<>());
        
        if (matchingRestCalls.isEmpty() && endpoint.buildComparableEndpointUri().endsWith("*")) {
            for (String uri : restCallMap.keySet()) {
                if (uri.startsWith(endpoint.buildComparableEndpointUri().substring(0, endpoint.buildComparableEndpointUri().length() - 1))
                    && endpoint.getHttpMethod().toLowerCase().equals(restCallMap.get(uri).get(0).getMethod().toLowerCase())) {
                    matchingRestCalls.addAll(restCallMap.get(uri));
                }
            }
        }
        
        if (matchingRestCalls.isEmpty()) {
            unusedEndpoints.add(endpoint);
        } else {
            endpointsAndMatchingRestCalls.add(new UsedEndpoints(endpoint, matchingRestCalls, endpointClass.getFilePath()));
        }
    }
}
```

Otherwise parallel streams could also do the work 😅

Reply:
I implemented our suggestion and did something similar for the REST Call analysis to reduce the number of nested loops there as well. Thanks for the suggestion!